src/c++/perf_analyzer/client_backend/openai/http_client.cc

```diff
@@ -0,0 +1,267 @@
+      return pem_str;
+    case HttpSslOptions::CERTTYPE::CERT_DER:
+      return der_str;
+  }
+  throw std::runtime_error(
+      "Unexpected SSL certificate type encountered. Only PEM and DER are "
+      "supported.");
+}
+
+const std::string&
+HttpClient::ParseSslKeyType(HttpSslOptions::KEYTYPE key_type)
+{
+  static std::string pem_str{"PEM"};
+  static std::string der_str{"DER"};
+  switch (key_type) {
+    case HttpSslOptions::KEYTYPE::KEY_PEM:
+      return pem_str;
+    case HttpSslOptions::KEYTYPE::KEY_DER:
+      return der_str;
+  }
+  throw std::runtime_error(
+      "unsupported SSL key type encountered. Only PEM and DER are "
+      "supported.");
+}
+
+void
+HttpClient::SetSSLCurlOptions(CURL* curl_handle)
+{
+  curl_easy_setopt(
+      curl_handle, CURLOPT_SSL_VERIFYPEER, ssl_options_.verify_peer);
+  curl_easy_setopt(
+      curl_handle, CURLOPT_SSL_VERIFYHOST, ssl_options_.verify_host);
+  if (!ssl_options_.ca_info.empty()) {
+    curl_easy_setopt(curl_handle, CURLOPT_CAINFO, ssl_options_.ca_info.c_str());
+  }
+  const auto& curl_cert_type = ParseSslCertType(ssl_options_.cert_type);
+  curl_easy_setopt(curl_handle, CURLOPT_SSLCERTTYPE, curl_cert_type.c_str());
+  if (!ssl_options_.cert.empty()) {
+    curl_easy_setopt(curl_handle, CURLOPT_SSLCERT, ssl_options_.cert.c_str());
+  }
+  const auto& curl_key_type = ParseSslKeyType(ssl_options_.key_type);
+  curl_easy_setopt(curl_handle, CURLOPT_SSLKEYTYPE, curl_key_type.c_str());
+  if (!ssl_options_.key.empty()) {
+    curl_easy_setopt(curl_handle, CURLOPT_SSLKEY, ssl_options_.key.c_str());
+  }
+}
+
+void
+HttpClient::Send(CURL* handle, std::unique_ptr<HttpRequest>&& request)
+{
+  std::lock_guard<std::mutex> lock(mutex_);
+
+  auto insert_result = ongoing_async_requests_.emplace(
+      std::make_pair(reinterpret_cast<uintptr_t>(handle), std::move(request)));
+  if (!insert_result.second) {
+    curl_easy_cleanup(handle);
+    throw std::runtime_error(
+        "Failed to insert new asynchronous request context.");
+  }
+  curl_multi_add_handle(multi_handle_, handle);
+  cv_.notify_all();
+}
+
+void
+HttpClient::AsyncTransfer()
+{
+  int place_holder = 0;
+  CURLMsg* msg = nullptr;
+  do {
+    std::vector<std::unique_ptr<HttpRequest>> request_list;
+
+    // sleep if no work is available
+    std::unique_lock<std::mutex> lock(mutex_);
+    cv_.wait(lock, [this] {
+      if (this->exiting_) {
+        return true;
+      }
+      // wake up if an async request has been generated
+      return !this->ongoing_async_requests_.empty();
+    });
+
+    CURLMcode mc = curl_multi_perform(multi_handle_, &place_holder);
+    int numfds;
+    if (mc == CURLM_OK) {
+      // Wait for activity. If there are no descriptors in the multi_handle_
+      // then curl_multi_wait will return immediately
+      mc = curl_multi_wait(multi_handle_, NULL, 0, INT_MAX, &numfds);
+      if (mc == CURLM_OK) {
+        while ((msg = curl_multi_info_read(multi_handle_, &place_holder))) {
```

Comment:
Would this while loop ever iterate more than once? If so, can you describe that situation?

Reply:
@GuanLuo

Reply:
Yes. https://curl.se/libcurl/c/curl_multi_info_read.html